src/net.cpp

```diff
@@ -1398,6 +1403,17 @@ void CConnman::ThreadSocketHandler()
                 }
             }
         }
+        //
+        // Reduce number of connections, if needed
+        //
+        if (vNodesCopy.size() > (size_t)nMaxConnections)
```

Comment:
I think this comparison needs to change (or we need to change the minimum, or both) because the peers in the set that we evaluate the size of here is larger than the set of evictable peers per `AttemptToEvictConnection()`, [causing this condition to trigger A LOT](https://user-images.githubusercontent.com/1410115/144773674-0db6197e-6e47-4239-b670-f56404fd414f.png), in cases that we set a low minimum (<28), i.e. have 8 outbound, and have less than the protected set (28 connections in total.)

Eviction skips over the following:
- [whitelisted peers](https://github.com/dogecoin/dogecoin/blob/31afd133119dd2e15862d46530cb99424cf564b0/src/net.cpp#L941)
- [outbound peers](https://github.com/dogecoin/dogecoin/blob/31afd133119dd2e15862d46530cb99424cf564b0/src/net.cpp#L943)
- [peers that are disconnected but not cleaned up yet](https://github.com/dogecoin/dogecoin/blob/31afd133119dd2e15862d46530cb99424cf564b0/src/net.cpp#L945)
- [4 random peers](https://github.com/dogecoin/dogecoin/blob/31afd133119dd2e15862d46530cb99424cf564b0/src/net.cpp#L959-L962)
- [8 peers with lowest ping](https://github.com/dogecoin/dogecoin/blob/31afd133119dd2e15862d46530cb99424cf564b0/src/net.cpp#L966-L969)
- [4 peers that most recently sent us tx](https://github.com/dogecoin/dogecoin/blob/31afd133119dd2e15862d46530cb99424cf564b0/src/net.cpp#L973-L976)
- [4 peers that most recently sent us blocks](https://github.com/dogecoin/dogecoin/blob/31afd133119dd2e15862d46530cb99424cf564b0/src/net.cpp#L980-L983)

I think that we can probably do the following:
1. Define the number of inbound peers that are protected from eviction => `PROTECTED_INBOUND_PEERS = 4 + 8 + 4 + 4` - in `net.h`
2. Change the condition to be `vNodesCopy.size() - std::max(MAX_OUTBOUND_CONNECTIONS, MAX_ADDNODE_CONNECTIONS) - PROTECTED_INBOUND_PEERS > nMaxConnections`.
3. Count the number of whitelisted peers... maybe `std::find_if` over `vNodesCopy`, but inside this condition?

Reply:
I've done 1 and 2. How does that work so far?

I don't like the cast necessary to make 2 behave properly, but making more of these values unsigned later will let me get rid of that.

I suppose 3 would be a net benefit, but repeatedly traversing `vNodesCopy` feels awkward to me. Maybe it's not so bad in practice, depending on how often people reduce their connection limits here.

Reply:
I pushed what might be a better solution. It adds a tiny bit of bookkeeping to every loop through this method, but it makes the intent of the code a lot clearer.

Reply:
Thanks! Not done testing the previous code yet (takes a while to get > 32 incoming peers even when you have 28 nodes yourself 🤔) but will move on to this as soon as I'm done with that.